### DISCUSSION TITLE
Equations overlapping with equation numbers

### DISCUSSION THREAD

**Discussion by pgeipi10** [KaTeX/KaTeX#3187] [Q&A] [answered]
Are there any good strategies for avoiding equations overlapping with equation numbers, as in the attached image? Some strategies that occur to me:
1. Shrink the font when this overlap is detected.
2. Display the equation number below the equation.
3. Use some css trick to allow the span to bust through the width of the parent container.
4. Force the equation to right align and therefore extend past the left margin.
<img width="1190" alt="Screen Shot 2021-08-25 at 1 35 39 PM" src="https://user-images.githubusercontent.com/7796257/130784705-a899bc07-f843-48d2-9b66-f20d006f5033.png">

**Comment by edemaine** [comment]
See #3066 for related discussion on fixing this in KaTeX proper (still lacking a solution).

If you don't care about proper centering, you could use the following CSS workaround:
```css
.katex-display>.katex>.katex-html>.tag { position: static; float: right; }
```

**Comment by pgeipi10** [comment, marked as answer]
I thought I didn't care!.. - until I saw the overall effect:

<img width="1312" alt="Screen Shot 2021-08-25 at 6 30 56 PM" src="https://user-images.githubusercontent.com/7796257/130829568-fe241883-50b5-43ab-81a9-dae938d8fbad.png">

**Reply by edemaine** [reply]
Oops, try adding `float: right` so at least the tags are right justified.

**Reply by pgeipi10** [reply]
Perfect now! Thanks!

**Reply by pgeipi10** [reply]
Coming back to this, your suggestion is a *vast* improvement over the default.
<img width="1208" alt="Screen Shot 2021-09-01 at 3 12 58 PM" src="https://user-images.githubusercontent.com/7796257/131730523-aa6460b7-388a-4493-98a6-828728e0c39d.png">

However, in the spirit of greed, is it possible to either
A. Have the too-long equation be pushed out to the left as opposed to pushing the equation number down,     or
B. Have the equation auto-shrink to fit within appropriate space?

I would be happy with either a CSS or JS solution.

**Reply by edemaine** [reply]
For A, it seems like adding the following successfully makes the overflow go left instead of right (based on [this StackOverflow question](https://stackoverflow.com/questions/218065/overflow-to-left-instead-of-right)), while centering when there isn't overflow:
```css
.katex-display { float: right; width: 100%; }
```

B should be possible in JavaScript by looking for `.katex` elements whose `scrollWidth` is larger than its `clientWidth` (I think), and then applying a CSS `transform: scale(...)`.  Without modification, this will scale the tag too though.

Alternatively, you might consider C: make overflowing equation scrollable horizontally. [This documentation](https://katex.org/docs/issues.html#css-customization) shows how to do that.

**Reply by pgeipi10** [reply]
C is a good suggestion and I'll explore it.

For A, I think a little more is needed to make it work properly with the equation tag. Here's what I'm getting:
<img width="1079" alt="Screen Shot 2021-09-01 at 5 42 42 PM" src="https://user-images.githubusercontent.com/7796257/131749636-b482611f-70d4-4b93-8dc5-f04ebc4f4a2e.png">

**Reply by edemaine** [reply]
I'm not sure what would cause that. Could you make a CodePen or similar to demo what produces that output, so that it's easier to play with CSS variations?

**Reply by pgeipi10** [reply]
Hey, I don't want it to look like I'm tricking you into debugging my code, but perhaps it would be easier if I just gave you a link to the page that shows this issue:

[https://grinfeld.org/books/IntroductionToTensorCalculus/Chapter19.html#Section_19_4](https://grinfeld.org/books/IntroductionToTensorCalculus/Chapter19.html#Section_19_4)

**Reply by edemaine** [reply]
Thanks, that helps. I see the issue now: KaTeX manually sizes the `\tag` to have the same height and depth as the equation, so that it has the correct vertical alignment when they are side-by-side.  The `float: right` of the tag moves the tag down, but it keeps the same vertical size.

This seems to be a general problem with the tag float hack; I don't see how to easily fix it. 🙁 Maybe with JavaScript there's a way to detect that the `.tag` floated / is on its own line, and then `display: none` the inner `.strut` in that case (and update on `resize` events).

**Reply by pgeipi10** [reply]
So that makes sense. But why isn't it pushing the whole equation out to the left as we intended?

**Reply by edemaine** [reply]
Ah, I see.  Replacing `width: 100%` with `min-width: 100%` seems to work.  Makes sense; allows it to grow beyond 100%.  (This rule is needed to make centering work when the equation doesn't fill the line.)

**Reply by edemaine** [reply]
@pgeipi10 By the way, I notice that initial page load is slow. Do you know how much of that is because of KaTeX? You might consider pre-rendering the math to HTML, instead of rendering it on page load. (Maybe best to open a separate thread about that. To advise, I'd need to know what build process you're using.)

**Reply by pgeipi10** [reply]
Yes, I believe it's KaTeX rendering. What does "build process" refer to?
(If you mean how did I generate HTML - I wrote a barebones latex->html converter in Java.)

**Reply by edemaine** [reply]
Yes, that's what I meant. To speed things up, in principle, you can find `$...$` and `$$...$$` and whatever other instances, and run the KaTeX conversion (e.g. via executing the [KaTeX cli](https://katex.org/docs/cli.html)), while running your LaTeX->HTML conversion process.  The HTML files will be bigger, so there's more network traffic I suppose, but the load time should be reduced I imagine.
